raft.go

```diff
@@ -308,7 +315,54 @@ func (r *Raft) runCandidate() {
 		case rpc := <-r.rpcCh:
 			r.mainThreadSaturation.working()
 			r.processRPC(rpc)
+		case vote := <-prevoteCh:
+			// This a pre-vote case but could lead to winning an election, in the case that majority of other nodes
+			// don't support pre-vote or have pre-vote deactivated.
+			r.mainThreadSaturation.working()
+			r.logger.Debug("got a prevote!!", "from", vote.voterID, "term", vote.Term, "tally", grantedVotes)
+			// Check if the term is greater than ours, bail
+			if vote.Term > r.getCurrentTerm() && !vote.PreVote {
+				r.logger.Debug("newer term discovered, fallback to follower", "term", vote.Term)
+				r.setState(Follower)
+				r.setCurrentTerm(vote.Term)
+				return
+			}
+			if vote.Term > term && vote.PreVote {
+				r.logger.Debug("newer term discovered on pre-vote, fallback to follower", "term", vote.Term)
+				r.setState(Follower)
+				r.setCurrentTerm(vote.Term)
+				return
+			}
+
```

Comment:
Do we need different logic for PreVote as well as this?

For example if the peer we sent it to denied us despite a higher term because our logs are behind, or because there is already a stable leader, we should treat that the same way and switch to Follower to try and wait for the leader to catch us up rather than just running pre vote over and over.

Actually I feel like that's not special to PreVote 🤔. Maybe we should do that for real elections too when we are rejected without a higher election term being present 🤔 .

Ideally, as a candidate, especially one having prevotes rejected, we need a path to transition back to follower so we can continue without disrupting the cluster. The only way out of candidate state right now I think is to either win an election in a new term ourselves, or observe another election happeneing - either way we can't become a healthy follower for the stable cluster leader without the cluster being disrupted by an election unless there is some other path to follower that _doesn't_ involve observing a term higher that our prevote one!

Reply:
I don't think this is needed in the election code, if we are part of the cluster already. The leader will eventually send an AppendEntry that will transition us to followers. Doing so in here have the potential, for example, of making a node not part of the cluster think it's a follower.